Implement future incompatibility report support

Aaron1011 opened the pull request:
cc rust-lang/rust#71249

This implements the Cargo side of 'Cargo report future-incompat'

Based on feedback from alexcrichton and est31, I'm implemented this a
flag `--future-compat-report` on `cargo check/build/rustc`, rather than
a separate `cargo describe-future-incompatibilities` command. This
allows us to avoid writing additional information to disk (beyond the
pre-existing recording of rustc command outputs).

This PR contains:

* Gating of all functionality behind `-Z report-future-incompat`.
  Without this flag, all user output is unchanged.
* Passing `-Z emit-future-incompat-report` to rustc when
  `-Z report-future-incompat` is enabled
* Parsing the rustc JSON future incompat report, and displaying it
  it a user-readable format.
* Emitting a warning at the end of a build if any crates had 
  future-incompat reports
* A `--future-incompat-report` flag, which shows the full report for 
  each affected crate.
* Tests for all of the above.

At the moment, we can use the `array_into_iter` to write a test.
However, we might eventually get to a point where rustc is not currently
emitting future-incompat reports for any lints. What would we want the 
cargo tests to do in this situation?

This functionality didn't require any significant internal changes to
Cargo, with one exception: we now process captured command output for 
all units, not just ones where we want to display warnings. This may 
result in a slightly longer time to run `cargo build/check/rustc` from
a full cache. since we do slightly more work for each upstream
dependency. Doing this seems unavoidable with the current architecture,
since we need to process captured command outputs to detect
any future-incompat-report messages that were emitted.

rust-highfive commented:
r? @Eh2406

(rust_highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
cc @ehuss @pnkfelix

Aaron1011 commented:
This can be tested on https://github.com/Aaron1011/arrayinto

Aaron1011 commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -913,6 +915,10 @@ fn build_base_args(
             .env("RUSTC_BOOTSTRAP", "1");
     }
 
+    if bcx.config.cli_unstable().future_incompat_report {
```
This check doesn't get taken into account by Cargo's caching - adding/removing this flag from `cargo build` will not cause anything to rebuilt. I'm unsure how big of an issue this is, since all of this will go away once things get stabilized.

ehuss commented:
Thanks!

I wouldn't worry about the caching of the `-Z` flag. I'm not expecting this to stay unstable for very long, but will give us a chance to test drive it.

As for errors, the answer is "it depends". Most of Cargo uses string-based errors with anyhow, and most errors are propagated with `?`. We also try to be careful to add context to every message (particularly for file I/O which often results in confusing errors without context). There is a [`chain_err`](https://github.com/rust-lang/cargo/blob/c46c06a1615912c207becdd8e1f393ff87d3fc08/src/cargo/util/errors.rs#L15) method to help with that (which should eventually be replaced with anyhow's `with_context` which is essentially the same).

I would recommend moving the code in the `command` module to the `ops` module. Generally the code in `command` is just arg parsing, and then the actual implementation is in an `ops` submodule. The `ops` side can return a `CargoResult` which is a little easier to work with than `CliResult`.

Since this is modifying the `JobQueue`, that area of Cargo is much more sensitive to managing errors since it must wait for all threads to finish even if there is an error. So, as you may have noticed there is some special handling of errors there.

It looks like you already found `drop_println!`, which deals with I/O errors.

As for tests, in the short term I think it would be fine just to pick a warning that is unlikely to change soon. An alternative would be to create a fake rustc that emits the appropriate JSON (the tests do that in several places like [this](https://github.com/rust-lang/cargo/blob/c46c06a1615912c207becdd8e1f393ff87d3fc08/tests/testsuite/cache_messages.rs#L401-L429). But that wouldn't be testing that it actually integrates with the real rustc properly. Another alternative would be to embed a fake lint in rustc directly, but I'm not sure if that's necessary.

As for handling canceled builds, it looks like you have support for that, which looks good to me. Is there a specific sequence of interactions that you are concerned about?

Let me know if you have any specific questions or particular areas you'd like me to look at. There's also the new [Contributor Guide](https://doc.crates.io/contrib/) which briefly provides some overviews of things like error handling. If there's something missing there, it may be useful to help highlight what needs expanding.

Aaron1011 commented:
> As for handling canceled builds, it looks like you have support for that, which looks good to me. Is there a specific sequence of interactions that you are concerned about?

If a user performed the following actions:
* Run a build to completion, which produces a future-incompat-report
* Change a bunch of things
* Start a new build, but immediately kill it before we opened and truncated the previous report
* Ran `cargo describe-future-incompatibilities`

Then 'stale' warnings would be shown. I'm not sure that this really matters - I would expect people who really care about these warnings to have the commands run on CI, where they would always run to completioin.

ehuss commented:
Yea, I wouldn't worry to much about that specific sequence.  We can always gauge later if people run into confusing situations or stale entries or whatever. It should be relatively easy to tweak and change this command.

When reading the file, it might be good to be somewhat tolerant of errors (print anything if possible, and then display an error if it has problems deserializing it). That might help in case the file gets partially written or corrupted or was generated with a different version of Cargo that changes the format, but even that I wouldn't worry too much about.

It's tough to predict every confusing circumstance, but here's one possibility:

1. `cargo test`: shows a warning for a dev-dependency
2. User editor runs `cargo check` in the background which will delete the file (and not generate a warning since it is not checking dev-dependencies).
3. User runs `cargo describe-future-incompatibilities` and it prints nothing.

I'm not sure how likely that is. I'm not too worried about it, but those are the kind of weird things that might confuse people.

Aaron1011 commented:
> User's editor runs cargo check in the background which will delete the file (and not generate a warning since it is not checking dev-dependencies).

Oh, that's quite subtle. I think it would be a good idea to add a header (or footer) to the displayed message to tell the user which cargo command generated the report.

est31 commented on src/bin/cargo/commands/describe_future_incompatibilities.rs:
```diff
@@ -0,0 +1,43 @@
+            FutureIncompatReportLine::Finished => has_end = true,
+        };
+    }
+    lines.sort_by(|r1, r2| r1.name.cmp(&r2.name));
```
Nit: you can do `sort_by_key` her

Aaron1011 replied:
That would require cloning the `name` value

est31 replied:
Good point, case rested.

est31 commented:
Nice work! I have only one minor criticism about the implementation, which is that it only prints the offending crate's name and never prints the version. The RFC [specifies](https://github.com/rust-lang/rfcs/commit/90415f83f07985f4e883ef2bf75a03666f943b8c) that cargo shall print the version if there are multiple packages with that name in the cargo build graph. I'd personally extend it from the build graph to the entire DAG encoded in Cargo.lock, basically using the same logic that adds version numbers to crates in `Cargo.lock` formats 2 onwards.

The reasoning [why I asked](https://github.com/rust-lang/rfcs/pull/2834#discussion_r357912488) for this to be added to the RFC was to make sure that users know which version is the offending one, so that they can base further actions on that.

From the maintainer POV of crates that yield such warnings, it might even be very useful to have the version number printed all the time, because the warning might not happen at the latest version of the crate but an older one. If the report includes the older unmaintained version number, the maintainer can immediately close the issue report and tell people to update (if that's their choice, maybe they consider the older version branch still maintained, then they want to publish an upgrade). If there is no version number in the report, the maintainer has to ask first which version of the crate creates the warning, or investigate themselves whether the latest version of their crate creates that warning, or whether it's an older version. It's more helpful to be told by the users though, because that's the version the users care about. Maybe it's 3 semver breaks ago, one never knows (and it doesn't help that the download graph of crates.io only lists a limited number of versions).

alexcrichton commented:
One of the reasons Cargo doesn't have the issues of reading files from old builds is that it's not something  that Cargo does today, basically for all the reasons mentioned here because it's quite hard. A case like @ehuss mentions [above](https://github.com/rust-lang/cargo/pull/8825#issuecomment-720760068) seems pretty worrisome to me, and could perhaps be solved by adding a CLI flag to `cargo build` (and other subcommands) rather than adding a new subcommand?

Aaron1011 commented:
@pnkfelix What are your thoughts on @alexcrichton 's proposal? Would this require amending the RFC?

est31 commented:
Actually that's a good point @alexcrichton . If the `cargo describe-future-incompatibilities` command would instead be a flag of `cargo build`/`cargo check` etc. it would be possible to reuse the cached json files that contain the output of rustc, thus removing the need for the extra level of caching that this PR adds. A bad alternative would be to make `cargo describe-future-incompatibilities` call `cargo build` etc. internally, but one doesn't know which precise command the user has put in previously, so one has to make a guess. The guess would be close but not precise though. Often it can lead to rebuilding of the entire crate graph only to give a result that maybe enables different cargo features etc. So a flag on `cargo build`/`cargo check` etc might be the way to go.

bors commented:
:umbrella: The latest upstream changes (presumably #8818) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Aaron1011 commented:
@alexcrichton @est31: I've re-implemented this as a `--future-incompat-report` flag on `cargo check/build/rustc`, as you suggested. In additional to eliminating the race condition issues, this also made the implementation significantly less complicated.

Aaron1011 commented:
@ehuss: This is now ready for review.

While I believe this still follows the spirit of [RFC 2834](https://rust-lang.github.io/rfcs/2834-cargo-report-future-incompat.html), the `describe-future-incompatibilities` command does not exist at all. Do we need to amend the RFC with this change - if so, what is the process for doing so?

ehuss commented:
Oh, sorry, I just saw you sent an update here.  I apologize, I did not post a comment sooner before you made some changes. We discussed this a little in our team meeting.  There were two things that we discussed:

The concern of the temporal nature of the cache:  If the cache is deleted and recreated each time a command is run, there is some concern that it will be confusing if it fluctuates (based on what is built), and the time lag between seeing the notification, and the user running the command to display the report (during which other commands may have been run).  I personally don't feel strongly about it, but would it be possible to change the on-disk representation to accommodate that concern?  Perhaps it could retain the last N runs (maybe as separate files, or all in one file), and the command would by default display the last report, but have an option to display a specific (older) report?  This is instead of using a flag as suggested above.  Would you be willing to maybe go in that direction?

The other thing we discussed was perhaps changing the top-level command name.  I feel a little uncomfortable adding a lot of highly specific top-level cargo commands, since it can make it a little harder to document and for users to navigate.  I was wondering what you think about changing it so that there is a `cargo report` command that takes a sub-command argument, like `cargo report future-incompatibilities`.  This would allow us to add additional report-like features, like "show build timings", or "explain why the last build wasn't cached", and they can be grouped within a single command.

I don't think in general that design tweaks need to go through any sort of process to amend the RFC. As long as the spirit of the original RFC is retained, I think it is up to each team to decide on the final implementation.

Aaron1011 commented:
>  Perhaps it could retain the last N runs (maybe as separate files, or all in one file), and the command would by default display the last report, but have an option to display a specific (older) report? This is instead of using a flag as suggested above. Would you be willing to maybe go in that direction?

Personally, I think this is very unlikely to be used:

* If I want to see the report for a specific command (e.g. `cargo check`), I think just adding `--future-incompat-report` will be generally faster than looking up a specific report in a list. Adding a flag is fast and easy to type, while viewing a list requires the user to find the particular command they are looking for in a potentially long list of commands.

* I think the two best use-cases of this feature are gating PRs on CI, and 'nudging' local users to update their dependencies when working on a crate locally. I don't think retaining a list of past reports is useful in either of those scenarios, and would require dealing with the various race conditions that @alexcrichton and @est31 were worried about.

alexcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -144,17 +146,16 @@ fn compile<'cfg>(
-            };
+            // We always replay the output cache,
+            // since it might contain future-incompat-report messages
+            let work = replay_output_cache(
```
This runs the risk of slowing down incremental compiles since this may do nontrivial work for all upstream crates (which can be a lot), can you try to do some testing locally to see what the impact is?

Aaron1011 replied:
I tested with `actix-web` (211 dependencies) - without running any kind of precise benchmarking, it looks like this may have increased the runtime of `cargo build` (with a full cache) by a few hundredths of a second (0.08 to 0.09 seconds before, 0.11 to 0.12 seconds after).

I think this worst case for this change would be:

1. Running `cargo build -vv` on a compilation graph with a significant number of normal upstream warnings.
2. Repeatedly running `cargo build` with the cache prepared by `cargo build -vv`, which cauess Cargo to process and discard a large number of cached diagnostic messages.

However, this seems like a pretty unlikely scenarion, and I haven't been able to measure any kind of significant impact locally (testing with `actix-web` again).

Eh2406 replied:
Drive by comment, what system was this testing on? Can we check on windows?

Aaron1011 replied:
I tested on Arch Linux

alexcrichton replied:
I agree that if we stick with this scheme that we'll want to test on Windows which typically has a slower filesystem than other platforms. Additionally the case I'm worried about is not when upstream compiles have output but rather they have no output so Cargo opens and closes a whole lot of files rapidly which can increase the latency of Cargo itself.

Aaron1011 replied:
Yes - we still want to display a note when running `cargo build` without any arguments.

alexcrichton replied:
This is another reason I would prefer to remove the flag on `cargo build` in the initial implementation. I don't believe this change is necessary if it's a separate subcommand, right? If so that would make the performance issue moot.

alexcrichton replied:
Hm ok I believe I'm wrong that this would change if there were only a subcommand. We'd still need to scrape files from past cached compilations to have the same report every time.

I don't think any testing of this has happened on Windows yet though?

ehuss commented:
The concern with a flag is that it may not be easy, or even possible to run cargo directly.  For example, if it is wrapped in another build system (like `x.py`), it may not be obvious which command to run, where to run it, which arguments to pass, etc.  Another concern is if someone is using an editor which uses cargo to drive messages, it also may be confusing to know how to recreate the same command on the command-line, and people sometimes have difficulty figuring out how to add additional cargo flags to the editor config.

In terms of the list, the intent is that the error message would specify exactly what you can run, maybe something like:

```
to see what the problems were, run `cargo report future-incompatibilities --report=7`
```

The user would copy and paste that, and see exactly which messages to display.  If they ran the command without any options, it could just display the last report.

I don't feel strongly about the difference, but I think the spirit of the RFC was to make it easy to view the report, regardless of what kind of environment cargo may be running in.

Aaron1011 commented:
@ehuss: It looks like @alexcrichton and @est31 both expressed a preference for a flag over a command. Is there any more work for me to do at the moment, or should I wait for a decision to be made?

est31 commented:
What about printing the path containing the report? Then you can just `cat` it.

Aaron1011 commented:
I don't write out the report to disk in the new implementation - it gets constructed in-memory by processing the (potentially recorded) `rustc` output lines.

est31 commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1114,6 +1119,13 @@ struct OutputOptions {
+    /// (unless we see a malformed message), discarding all other
+    /// messages. This is used when replaying the output of units
+    /// that we don't want to show warnings for, so that
+    /// decreasing the verbosity level doesn't caues us to
```
```suggestion
    /// decreasing the verbosity level doesn't cause us to
```

est31 commented:
@ehuss 

> For example, if it is wrapped in another build system (like x.py), it may not be obvious which command to run, where to run it, which arguments to pass, etc.

One could argue that a flag is better for `x.py` because `cargo generate-future-incompatibility-report` would look inside the target directory, but rust's buildsystem would most likely put the report into somewhere inside `build/`.

alexcrichton commented:
Sorry but to elaborate on my comment from earlier, I wanted to point out my concerns with the strategy of "use the last file to issue the report" with races and such. One solution is a flag to the build command but another solution is as @ehuss mentioned which is to have some sort of identifier which can uniquely identify one report.

After discussing I personally lean more towards keeping it as a separate command. We've gotten a lot of cargo reports recently about how configuring the build command line options or environment is difficult due to the build invocation being buried in a different system somewhere. This is leading us to lean more heavily on configuration files which is an out-of-band way to configure Cargo's build such that everyone invoking doesn't have to implement their own solution. This is a case which I think falls into that where having a separate command to read the report I think makes the most sense at this point.

bors commented:
:umbrella: The latest upstream changes (presumably #8837) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Aaron1011 commented:
Before I re-add the separate command (with numbered reports), does everyone agree that this is the route we want to take?

est31 commented:
Personally I think it'd be best for this separate command to take in a full path instead of an identifier, because those environments mentioned might move the target dir (e.g. by setting `CARGO_TARGET_DIR`) to locations that default cargo doesn't expect (x.py does this for example). And if it takes the full path, is it needed at all? Can the command offer significant features that go beyond cat? I lean towards yes, because it can detect whether the terminal supports ANSI colors or not, but that's about it.

Aaron1011 commented:
Printing a full path would make this more difficult to use on CI - you would either need to extract the path from the Cargo output, or hard-code what you think it should be. A flag doesn't have those kinds of issues.

est31 commented:
@Aaron1011 that's a good point I guess. Maybe one can have both a flag/.cargo config setting and a command?

Aaron1011 commented:
I've rebased this against master.

I think having some version of this available on nightly would be useful in resolving questions about a flag vs command. Everything is gated behind `-Z future-incompat-report`, so we aren't comitting to anything.

Aaron1011 commented:
Should I implement a separate command in addition to a flag, or can we land this with just a flag for experimentation?

alexcrichton commented:
If you'd like to land this immediately, then I would like to see the [performance question](https://github.com/rust-lang/cargo/pull/8825#pullrequestreview-526431662) resolved first, primarily for Windows. I don't think that's been followed up on. This is adding what has the possibility of being a nontrivial amount of work for incremental compiles of projects with many crates on, probably worst-so on Windows.

I am personally leaning towards a subcommand myself, but I'm not sure I understand the point about running on CI? How are users expected to run this on CI?

est31 commented:
@alexcrichton How I interpreted that remark was: if you get a warning on CI how do you get the detailed message? You can add something like `cargo print-future-incompat` to the ci files, but if there is the *requirement* for randomly generated keys/ids or one needs to cat a file at a special location whose path contains a changing hash, then you can't really add that to the files unless you make it extract the path/id from the stdout which is weird. You need something that works all the time. That's where a flag would be useful. Or a .cargo/config setting or something like that.

Note that CI environments can compile different code from the one the local dev compiles. One example is platform specific dependencies. So debugging it outside of CI might not be really easy.

bors commented:
:umbrella: The latest upstream changes (presumably #8441) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

ehuss commented:
I ran some performance tests on Windows, and didn't see any significant issues (it adds maybe 1-5 ms with caches warm). That makes sense to me, because in the normal case there are no output files, so the only cost is to try to open a non-existent file for each unit, which should be pretty fast.

I still feel like I would prefer a separate command as opposed to a flag.  I've seen people have difficulty figuring out how to wire Cargo flags into editors or larger build systems, or get confused about how different commands build different sets of dependencies or with different features.  However, I don't feel strongly about it, since I don't expect these warnings to come up often in practice.  I can help with implementing it if you'd like.  We could also move forward with a flag and add a separate command at a later date if it seems like it is too difficult for people to use.  @alexcrichton Do you have a preference one way or the other?

alexcrichton commented:
I don't have a preference beyond being able to reliably match the output of the report with the invocation that produced the report, so either is fine by me.

bors commented:
:umbrella: The latest upstream changes (presumably #8934) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Aaron1011 commented:
@alexcrichton Rebased against master

alexcrichton commented:
Ok thanks! We discussed this in a recent Cargo team meeting and concluded that for now a separate subcommand is the best option, however, rather than a flag-on-build. Would you be up for implementing that? The difference from before, however, is that an argument will indicate a report identifier of some kind to emit which is checked against the on-disk state to ensure it matches.

Aaron1011 commented:
@alexcrichton: Is the report id required? If so, it seems somewhat cumbersome to require users to make up an id when they invoke the command.

alexcrichton commented:
I personally believe that it's important to solve the problem where if you get told you should get the future incompatibility report that when you run the command you should be guaranteed to get *that* report, not something created in the meantime. My assumptions is that Cargo will print a copy/pasteable command which would include the Cargo-generated ID in it.

jyn514 commented:
>  For example, if it is wrapped in another build system (like x.py), it may not be obvious which command to run, where to run it, which arguments to pass, etc. 

If you make it a flag you could use `CARGOFLAGS` to pass arguments to cargo. Making it a separate subcommand means you'd have to modify bootstrap itself.

est31 commented:
I think it shouldn't be an either or, but a both. In some scenarios a flag or config setting are easier, in others a command is. Also the more methods there are, the less restrictive the feature feels to people.

Aaron1011 commented:
@alexcrichton Does having both a flag and a command sound reasonable to you?

alexcrichton commented:
If it's a small handful of lines to have both that doesn't seem unreasonable, but I would otherwise imagine that we wouldn't want too many ways to do the same thing. It can be a burden in Cargo to support if we're not careful.

Aaron1011 commented:
I've added back the `cargo describe-future-incompatibilities` command. It's invoked like this:

`cargo describe-future-incompatibilities --id <ID>`

The ID can be obtained from the last invocation of a cargo command (e.g. `cargo build`). Currently, the id is generated via the UUID crate (which I added as a dependency), but I'm open to alternatives.

If the provided id does not match the id stored in the file, the report is still printed (with a warning displayed before and after it). This should avoid frustrating situations where a user knows a report exists, but cannot access it.

Aaron1011 commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -764,6 +797,71 @@ impl<'cfg> DrainState<'cfg> {
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                drop(cx.bcx.config.shell().warn(&format!("the following crates contain code that will be rejected by a future version of Rust: {}",
```
I'm not sure how to get this into `full_report` while retaining the warning color.

Aaron1011 commented:
@alexcrichton: Are there any changes that you'd like me to make>

bors commented:
:umbrella: The latest upstream changes (presumably #8037) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

alexcrichton commented on tests/testsuite/future_incompat_report.rs:
```diff
@@ -0,0 +1,151 @@
+        .build();
+
+    for command in &["build", "check", "rustc"] {
+        p.cargo(&format!("{} -Z future-incompat-report", command))
```
Instead of `format!` you can do `.cargo(command).arg("-Zfuture-...")`

ehuss replied:
It seems like this was marked resolved, but the code looks the same.

alexcrichton reviewed (commented):
Thanks for the update and sorry for the delayed review, I've left a number of comments throughout.

Aaron1011 commented on src/cargo/util/command_prelude.rs:
```diff
@@ -212,6 +212,13 @@ pub trait AppExt: Sized {
+            "future-incompat-report",
+            "Ouputs a future incompatibility report at the end of the build (unstable)",
+        ))
+    }
```
This is the flag that I mentioned in https://github.com/rust-lang/cargo/pull/8825#issuecomment-751853641. It provides a way to print the report from a script (e.g. on CI) without needing to extract the generated ID from the human-readable output.

alexcrichton replied:
I feel like the infrastructure supporting both a flag and a subcommand is a bit larger than expected. If we want to cater to the CI use case we could make the `--id` argument optional to `cargo describe-future-incompat`, and if it's omitted we'd print the ID anyway to cross-reference in CI logs.

Aaron1011 replied:
Doing that just re-introduces the problem of getting the wrong report. By passing `--future-incompat-report`, you're guaranteed to see the correct report (since you specified the flag alongside the original command).

alexcrichton replied:
The biggest worry of that is concurrent processes like the RLS or deeply-nested invocations in other build systems. For CI, however, Rust projects are likely going to execute one command followed by another, so I don't think the original problem will come back. If anyone's watching CI they can always execute the same command locally if things differ.

Aaron1011 replied:
> If anyone's watching CI they can always execute the same command locally if things differ.

The report might contain lints from platform-specific code (e.g. the `winapi` crate), or the actual command being executed might be deep within various CI config files. I've struggled to reproduce CI failures before due to issues like this.

The vast majority of the code in this PR is either independent of the command/flag code, or exists to support writing the report to disk (for the command). Removing the flag would only eliminate a small amount of code, while re-introducing the very problem that the command was supposed to solve.

Aaron1011 replied:
> If anyone's watching CI they can always execute the same command locally if things differ.

It will be very difficult to tell if the command was overwritten. The user will need to find the ID printed out by `cargo describe-future-incompat`, and then look through the CI log to see if there was a command that printed out a *different* ID, which was executed *after* the command which printed out the *correct* ID. In practice, I can't imagine anyone doing this, so an overwritten report will likely go unnoticed.

Having a flag solves this issue completely. It also makes it easier to integrate into an existing CI setup, as there's no need to add a separate command invocation.

Aaron1011 replied:
Another advantage to the flag - CI may test a crate with different combinations of features, which may in turn enable different combinations of dependencies. I could easily see someone doing something like:

```
cargo test --features A
cargo test --features B
cargo describe-future-incompat
```

without realizing that this will not show a report for `cargo test --features A`. With a flag, it's immediately clear that `cargo test --features B --future-incompat-report` will only show a report for that specific command, and will *not* show a report for any other commands.

alexcrichton replied:
Personally I don't really understand the CI use case of this feature. These are all just warnings which don't fail a CI run and I don't know many folks that comb successful CI logs to see if there are any extra warnings or things like that. I'm also under the assumption that we can remove a lot of the support in this PR if there's only a subcommand. There seems like there's a good amount of gymnastics shuffling around whether or not warnings are printed, whether it's just future incompat, etc. It seems like what should happen is that we catch any future incompat message from the compiler, don't print it during `cargo build`, and save it to disk at the end of the build with a `--id` that can be used to regurgitate them later.

Again I understand the need for `--id`, I'm the one who brought it up. The main thing I'm worried about is concurrent Cargo invocations, not serial ones with different features/platforms/etc. If someone really cares about CI and is really combing all of their successful logs, then having a `describe-future-incompat` after each `cargo` command doesn't seem too bad.

Aaron1011 replied:
> These are all just warnings which don't fail a CI run and I don't know many folks that comb successful CI logs to see if there are any extra warnings or things like that

My intention was to add a way of doing this in an automated way in a follow-up PR (e.g. allow cargo to fail the build when `--future-incompat-report` is non-empty, or have `describe-future-incompat` set a non-zero exit code when the report is non-empty).

From my perspective, the CI use case is the most valuable part of this feature. The idea of this feature is to give users advance notice of when their crates will break. For a larger crate, a user running it locally will likely not test all of the dependencies (for example, a Linux user will never end up building a `winapi` dependency). What a crate author does with this information is up to them (e.g. an allow-fail CI job, or some kind of notification), but I think it's important that we provide the necessary tools for this to be set up. 

> The main thing I'm worried about is concurrent Cargo invocations

Has this come up in practice before for other commands? The two scenarios I can think of are:

1. Some kind of strange build system wrapper that spawns multiple simultaneous Cargo invocations for the same crate
2. A user having some kind of auto-build running (e.g. in their editor), which ends up running at the same time as an explicit `cargo build`.

For either of these scenarios to actually cause problems, they will need to either compile different sets of upstream dependencies, or set different features for upstream dependencies. Simply building the same upstream dependencies multiple times will result in the same report being generated, so it wouldn't which one actually got displayed.

I'm fine with supporting `--id` to try to prevent this, but I'm concerned that an important usecase (detecting future breakage on other platforms which are tested by CI) is being detrimentally affected by a very obscure one.

Aaron1011 replied:
>  There seems like there's a good amount of gymnastics shuffling around whether or not warnings are printed, whether it's just future incompat, etc.

See https://github.com/rust-lang/cargo/pull/8825#discussion_r564666799. This logic is needed to support correctly generating a report when using cached upstream dependencies, and would not go away if we just had the command.

Aaron1011 replied:
> If someone really cares about CI and is really combing all of their successful logs, then having a describe-future-incompat after each cargo command doesn't seem too bad.

They *can*, but I think people might expect the command to print some kind of summary of all commands, instead of the most recent one. Doing:

```rust
cargo build --features A
cargo build --features B
cargo describe-future-incompat
```

will *appear* to work, but will miss any future-incompat messages that only show with with feature 'A'.

alexcrichton replied:
Sorry this PR grows ever longer with comments and I feel like the same things are being said. I'm not really sure that you're understanding me. Every time I come back to this I have to boot up on ever-more comments about what is going on here and it feels like I'm reading the same thing every time.

I'm saying we need `--id`. I'm ok with adding `cargo build --future-incompat-report` for now as well. Is there anything else here that needs to be discussed?

Aaron1011 commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1347,13 +1369,21 @@ fn replay_output_cache(
+        // future incompat report messages, not any other diagnostics
+        // (which would be suppressed by `--cap-lints` if we were
+        // doing a full rebuild).
+        replay_future_incompat_only: !show_warnings,
```
Previously, we gated replaying messages on `unit.show_warnings`. We now need to replay messages for all units in order to detect future-incompat-report messages. However, we need to respect `show_warnings`, and avoid replaying warning messages for crates that we don't want to show warnings for.

alexcrichton replied:
Do warnings get printed if this is removed? The `--cap-lints` flag should prevent the compiler from generating lints most of the time here I thought.

Aaron1011 replied:
The problem is when we have a sequence of commands like this:

1. `cargo build --verbose` (we don't pass `--cap-lints`, causing us to cache warning messages)
2. `cargo build` (we replay the cached messages, which includes warnings that we no longer want to show).

Previously, we would not replay any messages for the second `cargo build` invocation (due to `show_warnings` being `false`), so it didn't matter that we cached them.

alexcrichton replied:
Hm ok I forgot that we tried to handle that. This is the support I was thinking of though where I would prefer to leave it to a separate subcommand if we can rather than editing lots of stuff in the main compile pipeline to do the right gymnastics with figuring out where warnings go.

Aaron1011 replied:
All of this logic is independent of the way in which we display the output (flag vs subcommand). If the user does:

1. `cargo build`
2. edit a file in the local crate
3. `cargo build`
4. Display the future incompat report (either through a command or flag)

Then the second `cargo build` invocation needs to replay cached messages from upstream crates in order to generate the correct report.

Actually, supporting *only* a flag made the implementation significantly less complicated, since  we no longer need to serialize anything to disk. Once we have the requirement of serializing this to disk (in order to support a command), the actual flag implementation is trivial - we just [need to check it before printing the report](https://github.com/Aaron1011/cargo/blob/3199de8b5de6f440cdb6e7a3c8401f351afa7344/src/cargo/core/compiler/job_queue.rs#L854-L863)

Aaron1011 commented on src/bin/cargo/commands/describe_future_incompatibilities.rs:
```diff
@@ -0,0 +1,68 @@
+    };
+
+    if bad_id {
+        warning_message = Some(format!("Expected an id of `{}`, but `{}` was provided on the command line. Your report may have been overwritten by a different one.", on_disk_report.uuid, raw_id));
```
In my experience, it can be extremely frustrating when a command refuses to display information that I know it has access to. I think it's better to err on the side of showing more information rather than less.

alexcrichton replied:
The purpose of the `--id` was to handle a case where you get one set of warnings but then in the interim a separate set of warnings is generated. The whole purpose of `--id` is to get the guaranteed same set of warnings from a previous build. I understand your frustration, but in my mind getting *misleading* informatiton trumps that concern and means we should not display anything here if the wrong `--id` is passed.

Aaron1011 replied:
I've changed the code to bail out if a bad 'id' is specified.

Aaron1011 commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -764,6 +797,71 @@ impl<'cfg> DrainState<'cfg> {
+                if self.show_future_incompat_report {
+                    drop_eprint!(cx.bcx.config, "{}", full_report);
+                    drop(cx.bcx.config.shell().note(
+                        &format!("this report can be shown with `cargo describe-future-incompatibilities -Z future-incompat-report --id {}`", uuid)
```
Should I keep the `-Z` flag for gating `--future-incompat-report`, or get rid of it entirely?

alexcrichton replied:
For a brand new subcommand I think it's fine to just say that the subcommand works on nightly and otherwise unconditionally prints an error on stable/beta. I'm not sure that we have a ton of precedent for new unstable subcommands.

Aaron1011 replied:
I meant for the flag pass to `cargo build`. If we don't have a `-Z` flag, then we'll effectively make this on-by-default on Nightly, since we'll start printing the short notice when running `cargo build`.

alexcrichton replied:
Yes if `cargo build` has a `--future-incompat-report` flag then it needs a `-Z` flag to gate that flag.

Aaron1011 commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -764,6 +797,71 @@ impl<'cfg> DrainState<'cfg> {
                 drop(cx.bcx.config.shell().status("Finished", message));
             }
+
+            if !self.future_incompat_reports.is_empty() {
```
I assume you mean the `-Z` flag? We want to run this logic regardless of whether or not the user wants the report displayed, so that we can save the report to disk and print a notice.

alexcrichton replied:
I mean this should be in the `if` block above.

Aaron1011 commented on src/cargo/util/command_prelude.rs:
```diff
@@ -466,6 +473,15 @@ pub trait ArgMatchesExt {
+                )
+            }
+            build_config.future_incompat_report = true;
+        }
```
We already have `-Z future-incompat-report` - using the same code as the previous opts will require `-Z unstable-opts` as well, which seems redundant.

alexcrichton replied:
I think it would be better to pattern this after the other options because `--future-incompat-report` is an unstable option and the unstable error message is more consistent with the above pattern.

Aaron1011 replied:
I think there's been some confusion about exactly which flags do what. This is my proposal:

1. The output of a plain `cargo build` on Nightly is unchanged - that is, we do *not* print any kind of message suggesting that the user display the full future-incompat-report, even if one was generated.
2. To display this message, the user needs to run `cargo build -Z future-incompat-report`, which enables the logic for generating the report, and notifying the user about it.
3. Since we already have a `-Z` flag, we can use it to gate the `--future-incompat-report` flag for displaying the report - e.g. `cargo build -Z future-incompat-report --future-incompat-report`.
4. Optionally, `cargo describe-future-incompat` can require `-Z future-incompat-report`, or can be nightly only.
If I understand your proposal correctly, we would need to either:

1. Use the flag as `cargo build -Z future-incompat-report -Z unstable-opts --future-incompat-report`, which seems incredibly verbose.
2. Use the flag as `cargo build -Z unstable-opts --future-incompat` report. If we allow `cargo describe-future-incompat` on nightly without any flags, then the only usage of `-Z future-incompat-report` is for displaying the notice (but not the full report) via `cargo build -Z future-incompat-report`.

My motivation here was to avoid changing the behavior of any existing commands without explicit opt-in. If you wanted, we could simplify things by having `cargo build` print the future-incompat notice on Nightly, without requiring any flags.

alexcrichton replied:
Yes I understand what this PR is doing, and I'm saying that this should be:

```
cargo build -Z unstable-opts --future-incompat-report
```

Cargo can internally recognize that a report is being generated and implicitly enable  `-Z future-incompat-report`. There's not really any reason to additionally require `-Z future-incompat-report`. I'm fine requiring it as well though if you'd prefer. Commands should indeed remain as they are unless new flags are passed.

Aaron1011 replied:
I've changed the code to require voth `-Z unstable-opts` and `-Z future-incompat-report`.

alexcrichton commented on src/cargo/core/features.rs:
```diff
@@ -364,6 +364,7 @@ pub struct CliUnstable {
     pub weak_dep_features: bool,
     pub extra_link_arg: bool,
     pub credential_process: bool,
+    pub future_incompat_report: bool,
```
I've gotten confused by how many different occurrences of "future incompat report" is used in various places. I think this should be renamed `enable_future_incompat_reporting` since this is just what enables the lints in rustc.

Aaron1011 replied:
I think `enable_future_incompat_feature` would be better, to emphasize that this doesn't necessarily enable displaying the report. I'll rename some of the other duplicate field names as well.

alexcrichton commented on src/cargo/core/compiler/mod.rs:
```diff
@@ -1135,6 +1140,13 @@ struct OutputOptions {
+    /// that we don't want to show warnings for, so that
+    /// decreasing the verbosity level doesn't caues us to
+    /// incorrectly replay warnings.
+    replay_future_incompat_only: bool,
```
I think this is probably better renamed as `show_warnings` like the original input for the option rather than making this future-incompat-specific. That's basically what this option is, which is that if we're unconditionally reading the output file of rustc then we just want to skip warnings, but other directives may wish to be processed.

Aaron1011 replied:
That's not what this flag does, though. If `report_future_incompat_only` is `true`, then we bail out of `on_stderr_line_inner` [if we fail to parse a future-incompat-report message](https://github.com/Aaron1011/cargo/blob/3199de8b5de6f440cdb6e7a3c8401f351afa7344/src/cargo/core/compiler/mod.rs#L1236-L1238), *before* attempting to do any other parsing.

I'm intentionally skipping the processing of other directives, because that's the current behavior of Cargo (we don't parse *any* directives, since we don't even read the file). We could put any future-incompat-report lines in their own file during recording of rustic output, but that seems much more complicated.

alexcrichton replied:
The change in this PR is to unconditionally on every single build *always* parse all historical outputs of rustc. We can't change that. I would prefer to not have tons of future-incompat-report related things all over Cargo since it does not seem like a feature that justifies so much weight. By renaming this to `show_warnings` it means that it's just whether or not upstream compiler warnings are shown. Future-incompat messages and all other output from rustc is always processed as usual, where we hide future-incompat, jobserver, etc.

Aaron1011 replied:
My concern was that this could have other side effects, particularly if we ever add any other types of recorded messages. However, if you'd like to limit the amount of logic specific to future-incompat stuff, I'll change this to `show_warnings`.

Aaron1011 commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -764,6 +797,71 @@ impl<'cfg> DrainState<'cfg> {
+                    // Writing to a String can never error
+                    writeln!(&mut full_report, "The crate `{}` v{} currently triggers the following future incompatibility lints:", report.krate, report.version).unwrap();
+                    for item in report.report {
+                        for line in item.diagnostic.rendered.lines() {
```
I copied over some of the `strip_ansi_escapes` logic from the stderr line processing. Do you think it's okay to output ANSI escapes to the file when the terminal supports colors, or should we require an explicit flag for that?

Aaron1011 commented:
@alexcrichton: I've addressed your comments

Aaron1011 commented:
@alexcrichton Are there any changes that you'd like me to make?

bors commented:
:umbrella: The latest upstream changes (presumably #9175) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

Aaron1011 commented:
Rebased

ehuss commented on src/bin/cargo/commands/describe_future_incompatibilities.rs:
```diff
@@ -0,0 +1,50 @@
+    report_file
+        .file()
+        .read_to_string(&mut file_contents)
+        .map_err(|e| anyhow!("Failed to read report: {:?}", e))?;
```
Generally the preferred method for Cargo error messages to use chaining to add context.  The error message format usually starts with a lowercase letter (similar to rustc).  Also, it is preferred to avoid displaying debug-formatted things.  This can be used with `use cargo::util::CargoResultExt;`.

```suggestion
        .chain_err(|| "failed to read report")?;
```

ehuss commented on src/bin/cargo/commands/describe_future_incompatibilities.rs:
```diff
@@ -0,0 +1,50 @@
+
+    let id = args.value_of("id").unwrap();
+    if id != on_disk_report.id {
+        return Err(anyhow!("Expected an id of `{}`, but `{}` was provided on the command line. Your report may have been overwritten by a different one.", on_disk_report.id, id).into());
```
Can you try to keep lines under 100 characters?  Breaks with backslash in the string can help.  Long lines like these break rustfmt.

ehuss commented on src/bin/cargo/commands/describe_future_incompatibilities.rs:
```diff
@@ -0,0 +1,50 @@
+        .file()
+        .read_to_string(&mut file_contents)
+        .map_err(|e| anyhow!("Failed to read report: {:?}", e))?;
+    let on_disk_report: OnDiskReport = serde_json::from_str(&file_contents).unwrap();
```
This unwrap doesn't seem safe. If the on-disk format changes in the future, it could fail to load.

```suggestion
    let on_disk_report: OnDiskReport = serde_json::from_str(&file_contents)
      .chain_err(|| "failed to load report")?;
```

ehuss commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -151,6 +157,8 @@ struct DrainState<'cfg> {
 
     // How many jobs we've finished
     finished: usize,
+    show_future_incompat_report: bool,
```
I'm not sure I see why `show_future_incompat_report` needs to be here. It's only used in one place, where `cx` is available, so maybe it could removed and just access the `build_config` value directly?

ehuss commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -583,6 +604,14 @@ impl<'cfg> DrainState<'cfg> {
                 }
             }
+            Message::FutureIncompatReport(id, report) => {
+                let unit = self.active[&id].clone();
+                self.per_crate_future_incompat_reports
+                    .push(FutureIncompatReportCrate {
+                        package_id: unit.pkg.package_id(),
```
I'm not sure I see why this needs to be cloned, maybe something like this?

```suggestion
                let package_id = self.active[&id].pkg.package_id();
                self.per_crate_future_incompat_reports
                    .push(FutureIncompatReportCrate {
                        package_id,
```

ehuss commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -763,14 +792,93 @@ impl<'cfg> DrainState<'cfg> {
+            let crates_and_versions = self
+                .per_crate_future_incompat_reports
+                .iter()
+                .map(|r| format!("{}", r.package_id))
```
```suggestion
                .map(|r| r.package_id.to_string())
```

ehuss commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -763,14 +792,93 @@ impl<'cfg> DrainState<'cfg> {
+                .collect::<Vec<_>>()
+                .join(", ");
+
+            drop(cx.bcx.config.shell().warn(&format!("the following crates contain code that will be rejected by a future version of Rust: {}",
```
Long line.

ehuss commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -763,14 +792,93 @@ impl<'cfg> DrainState<'cfg> {
+                .collect();
+
+            for report in std::mem::take(&mut self.per_crate_future_incompat_reports) {
+                full_report.push_str(&format!("The crate `{}` currently triggers the following future incompatibility lints:\n", report.package_id));
```
Long line.

ehuss commented on src/cargo/core/compiler/job_queue.rs:
```diff
@@ -763,14 +792,93 @@ impl<'cfg> DrainState<'cfg> {
+                })
+                .err();
+            if let Some(e) = err {
+                drop(cx.bcx.config.shell().warn(&format!(
+                    "Failed to open on-disk future incompat report: {:?}",
+                    e
+                )));
```
There's a function for doing this:

```suggestion
                crate::display_warning_with_error(
                    "failed to write on-disk future incompat report",
                    &e,
                    &mut cx.bcx.config.shell(),
                );
```

ehuss commented on tests/testsuite/future_incompat_report.rs:
```diff
@@ -0,0 +1,165 @@
+    for command in &["build", "check", "rustc"] {
+        p.cargo(&format!("{} -Z future-incompat-report", command))
+            .masquerade_as_nightly_cargo()
+            .with_stderr_does_not_contain("  = note: `#[warn(array_into_iter)]` on by default")
```
I'm pretty uncomfortable with all of these `with_stderr_does_not_contain` calls.  That function is pretty fragile, because any change in output means it will no longer be validating the behavior.  And rustc's output is prone to change frequently, and there will be no way to know when it changes.

Usually to make it more reliable, I would have a test that has `with_stderr_contains("[..]example[..]")`, and then a second test that validates the opposite with the exact same string with `with_stderr_does_not_contain("[..]example[..]")`.

And, in general, when checking rustc's output, I try to use wildcards as much as possible.  If someone wants to change the output, and it breaks cargo's testsuite, then it requires a hassle to update cargo to somehow work with the change.

In these cases, it might be less fragile to match against things like `"[..]array_into_iter[..]"`.

ehuss commented on src/cargo/core/compiler/future_incompat.rs:
```diff
@@ -0,0 +1,36 @@
+    pub diagnostic: Diagnostic,
+}
+
+/// A diagnostic emitted by the compiler a a JSON message.
```
Looks like a typo, I'm not sure what exactly it is trying to say.

ehuss reviewed (commented):
Can you also add some brief documentation to `unstable.md`?

ehuss commented:
I wanted to write a few more concerns about the implementation. I don't think any of these need to be addressed now, but I think they need to be considered eventually.

* The way color is handled seems a little awkward. One of the use cases is that the user runs cargo in an editor, sees the message that they need to run `cargo describe-future-incompatibilities`, they do that, and the message isn't colored. Not a big deal, color isn't super important, but that seems like a common situation. Ideally I would think the report data would be saved to the file in a more raw form, and then just re-render it as requested.
* This doesn't really work in editors (the user will never see the warning). It's a bit of an unfortunate behavior of Cargo in general (#8283), and the way most editor integration works. That seems like a concern, since in many cases that means the messages won't be seen.
* Requiring `-Z future-incompat-report` with `cargo build -Z unstable-opts --future-incompat-report` seems a bit awkward (it could be implied by the presence of `--future-incompat-report`). I understand why it is done that way, and hopefully it won't be too much of a problem.
* I think the command should save at least a few reports in the on-disk file, so that if there are things like background jobs (like `cargo watch`), it doesn't make it difficult to view a report. Per my suggestion far above, I think it would be easy to save them in an array, and just trim it to N reports. It could also use monotonic numeric IDs instead of the strange random IDs (the next ID could be computed by 1+ the max ID in the file). Then the report command could default to showing the report with the greatest ID when not given an ID.
* The message to "run cargo describe-future-incompatibilities" could also include the target directory if using a non-default one.
* The RFC discussed a mechanism to silence the warnings.  I'm not sure exactly what that should look like, but un-silenceable warnings can be frustrating, especially if it is not possible to update the offending dependency (or no new version is available).

Aaron1011 commented:
> I think the command should save at least a few reports in the on-disk file, so that if there are things like background jobs (like cargo watch), it doesn't make it difficult to view a report. Per my suggestion far above, I think it would be easy to save them in an array, and just trim it to N reports. It could also use monotonic numeric IDs instead of the strange random IDs (the next ID could be computed by 1+ the max ID in the file). Then the report command could default to showing the report with the greatest ID when not given an ID.

I'm concerned that this PR is adding an increasingly large amount of complexity to deal with a very unusual situation. See my previous comment https://github.com/rust-lang/cargo/pull/8825#discussion_r564675726 - can you come up with a concrete sequence of commands that will actually cause a report to get overwritten with a *different* one (as opposed to an identical version of the existing report)?

ehuss commented:
If I run `cargo test` in my terminal, then head over to my editor and make some changes while that works.  The editor runs `cargo check`, which does not include dev dependencies, and overwrites the report from `cargo test`. I had back over to my terminal and see the warning.  I run `cargo describe-future-incompatibilities --id=abcd`, and it returns the error: `Your report may have been overwritten by a different one.`.  Two things:

* That's an unpleasant experience. It does print the ID for the last report, but it seems confusing to me.
* The information for `cargo test` may be different from `cargo check` due to different dependencies.

Can you help me understand what the complexity would be?  I would expect it to just be a few lines of code. Something like:

```rust
let mut on_disk_report = load_reports()?;
if on_disk_report.reports.len() > MAX_REPORTS {
    on_disk_report.reports.remove(0);
}
let id = on_disk_report.reports.iter().max_by_key(|report| report.id).unwrap_or(0) + 1;
on_disk_report.reports.push(Report{id, contents});
on_disk_report.save()?;
```

I'm fine if you don't want to implement it here, it can be done later.

Aaron1011 commented:
@ehuss: I'd like to implement that later, if it's okay with you. I think it would be good to get *some* version of this available on nightly, so that people can start testing this out in various scenarios.

Aaron1011 commented:
@ehuss: I've addressed your review comments. Could you add your other implementation concerns to the rustc tracking issue?

bors commented:
:umbrella: The latest upstream changes (presumably #9022) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

ehuss commented:
Yep, I think addressing those later is fine.  I pushed a few minor changes and updated the tracking issue.

Thanks! 🎉 

@bors r+

bors commented:
:pushpin: Commit 139ed73f536d73467161cf2b88b2ed173ce52a99 has been approved by `ehuss`

<!-- @bors r=ehuss 139ed73f536d73467161cf2b88b2ed173ce52a99 -->
<!-- homu: {"type":"Approved","sha":"139ed73f536d73467161cf2b88b2ed173ce52a99","approver":"ehuss"} -->

bors commented:
:hourglass: Testing commit 139ed73f536d73467161cf2b88b2ed173ce52a99 with merge 7441e8c234ea60c0d07f7c1323c60e26b552e64e...
<!-- homu: {"type":"BuildStarted","head_sha":"139ed73f536d73467161cf2b88b2ed173ce52a99","merge_sha":"7441e8c234ea60c0d07f7c1323c60e26b552e64e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/2043491047)
Approved by: ehuss
Pushing 7441e8c234ea60c0d07f7c1323c60e26b552e64e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/2043491047"},"merge_sha":"7441e8c234ea60c0d07f7c1323c60e26b552e64e"} -->

LukasKalbertodt commented:
@Aaron1011 Thanks for your work on this! This feature will be a great tool for developing the standard library.
